### PR TITLE
cannon: Simplify load/stores with helper functions

### LINKED ISSUE #12562: cannon: Memory selection using LoadSubWord and StoreSubWord
Use the convenience functions for specific width memory access for arithmetic instructions.
Right now, instruction execution manually uses a lot of bitwise operations to select the appropriate bits for a given operation. This is hard to maintain and understand and should be simplified using the existing `LoadSubWord` and `StoreSubWord` utilities.

### PR DESCRIPTION
**Pull request by Inphi**
This patch reuses the same logic for the 32-bit and 64-bit VMs for subword access. This simplifies VM logic greatly and allows us to easily port the 32-bit VM to 64-bits in Solidity.

## Testing

- Existing 32-bit tests
- Manual run of `go test -tags=cannon64  -run 'TestEVMSingleStep_LoadStore64' -v ./mipsevm/tests`

## Meta
Builds on https://github.com/ethereum-optimism/optimism/pull/12483
Fixes https://github.com/ethereum-optimism/optimism/issues/12562

### CODE REVIEW

**Comment by Inphi** [comment]
I'll rebase this on develop later tonight

File: cannon/mipsevm/exec/mips_instructions.go [hunk 1]
```diff
@@ -343,48 +343,35 @@ func ExecuteMipsInstruction(insn uint32, opcode uint32, fun uint32, rs, rt, mem
 		case 0x0F: // lui
 			return SignExtend(rt<<16, 32)
 		case 0x20: // lb
-			msb := uint32(arch.WordSize - 8) // 24 for 32-bit and 56 for 64-bit
-			return SignExtend((mem>>(msb-uint32(rs&arch.ExtMask)*8))&0xFF, 8)
+			return SelectSubWord(rs, mem, 1, true)
```
Comment (mbaxter) [inline cannon/mipsevm/exec/mips_instructions.go:346]: Shouldn't we port these methods into `MIPSInstructions.sol`?

**Review by mbaxter — COMMENTED** [review]
Looks good - I just thought we would update the Solidity with these changes as well.

File: cannon/mipsevm/exec/mips_instructions.go (hunk 1 above)
Comment (Inphi) [reply cannon/mipsevm/exec/mips_instructions.go:346]: Yes. Will do.

Comment (Inphi) [reply cannon/mipsevm/exec/mips_instructions.go:346]: Done.

File: packages/contracts-bedrock/src/cannon/libraries/MIPSInstructions.sol [hunk 2]
```diff
@@ -667,4 +667,69 @@ library MIPSInstructions {
...
+        return subword_;
+    }
+
+    /// @notice Returns a word that has been updated by the specified subword at bit positions determined by the virtual
+    /// address
+    /// @param _vaddr The virtual address of the subword.
+    /// @param _memWord The full word to update.
+    /// @param _byteLength The size of the subword.
+    /// @param _value The subword that updates _memWord.
+    function updateSubWord(
```
Comment (mbaxter) [inline packages/contracts-bedrock/src/cannon/libraries/MIPSInstructions.sol:696]: Looks like the store ops still need to be updated to use this.

Comment (Inphi) [reply packages/contracts-bedrock/src/cannon/libraries/MIPSInstructions.sol:696]: Ack! Thanks for flagging

Comment (Inphi) [reply packages/contracts-bedrock/src/cannon/libraries/MIPSInstructions.sol:696]: Fixed.

File: packages/contracts-bedrock/src/cannon/libraries/MIPSInstructions.sol [hunk 3]
```diff
@@ -667,4 +663,69 @@ library MIPSInstructions {
...
+        uint32 _memWord,
+        uint32 _byteLength,
+        bool _signExtend
+    )
+        internal
+        pure
+        returns (uint32 subword_)
+    {
+        (uint32 dataMask, uint32 bitOffset, uint32 bitLength) = calculateSubWordMaskAndOffset(_vaddr, _byteLength);
+        subword_ = (_memWord >> bitOffset) & dataMask;
```
Comment (mbaxter) [inline packages/contracts-bedrock/src/cannon/libraries/MIPSInstructions.sol]: (nit) we're using `retVal` offchain
```suggestion
        retVal_ = (_memWord >> bitOffset) & dataMask;
```

File: cannon/mipsevm/exec/mips_instructions.go [hunk 4]
```diff
@@ -663,45 +628,56 @@ func HandleRd(cpu *mipsevm.CpuScalars, registers *[32]Word, storeReg Word, val W
...
 	mem := memory.GetWord(effAddr)
 
 	// Modify isolated sub-word within mem
-	dataMask, bitOffset, _ := calculateSubWordMaskAndOffset(addr, byteLength)
+	newMemVal := UpdateSubWord(vaddr, mem, byteLength, value)
+	memory.SetWord(effAddr, newMemVal)
+}
+
+// SelectSubWord selects a subword of byteLength size contained in memWord based on the low-order bits of vaddr
+// This is the nearest subword that is naturally aligned by the specifie dbyteLength
```
Comment (mbaxter) [inline cannon/mipsevm/exec/mips_instructions.go]: small typo:
```suggestion
// This is the nearest subword that is naturally aligned by the specified byteLength
```

**Review by mbaxter — APPROVED** [review]

**Pull request merged**